# Home task (Task about sum and product)

## 1. Добуток: "Я не знаю ці числа"

Це означає, що добуток точно не є унікальним, тобто таким, який утворює тільки якась одна пара чисел.

Наприклад якщо добуток - __21__, то він є ункіальним, оскільки існує лише одна пара чисел, яка його утворює __(3, 7)__ 

Якби добуток був унікальним, то можна було б зразу здогадатись що це за числа.

Також з цього можна зробити висновок, що обидва числа не можуть бути одночасно простими, оскільки добуток двох простих чисел є завжди унікальним.

## 2. Сума: "Я знаю, що ти не знаєш"

З цих слів можна зробити висновок, що число, яке сказали цьому вченому точно не є сумою двох простих чисел, або інших чисел, добуток яких є унікальним.

Цей вчений побачив, що зі всіх можливих пар чисел, які можуть складати дану йому суму, не можна утворити унікальний добуток. Відповідно він знав, що вчений, якому сказали добуток, не назве правильну пару чисел, оскільки добуток, який йому назвали не є унікальним і його можуть утворити більше ніж одна пара чисел.

In [1]:
MIN_VALUE = 1
MAX_VALUE = 100

def count_factors(product):
    count = 0
    for i in range(2, product//2):
        if product % i == 0:
            # Число не може бути більшим за 100 за умовою
            if (i < MAX_VALUE and product / i < MAX_VALUE):
                count += 1
    return count

def split_sum(sum):
    additions = []
    for i in range(2, sum//2+1):
        additions.append((i, sum - i))
    return additions    

In [2]:
def filter_sums(MIN_SUM, MAX_SUM):
    sums = []
    for i in range(MIN_SUM, MAX_SUM):
        count = 0
        for addition in split_sum(i):
            if count_factors(addition[0] * addition[1]) <= 2:
                count += 1

        if (count == 0):
            sums.append(i)
    return sums 


MIN_SUM = 4 # 2+2 (мінімальна можлива комбінація чисел)
MAX_SUM = 198 # 99+99 (максимальна можлива комбінація чисел)
print("Всі можливі суми: ")
sums = filter_sums(MIN_SUM, MAX_SUM)
print(sums)

Всі можливі суми: 
[11, 17, 23, 27, 29, 35, 37, 41, 47, 53]


Код вище можна оптимізувати. Виявляється, що якщо будь-яке просте число, яке є більшим за половину максимально допустимого числа в задачі (в нашому випадку це __100__), є одним із множників, то добуток буде унікальним числом. Саме тому всі суми, більші за __53__, відкидаються, оскільки число __53__
є простим і є більшим ніж половина числа __100__.

In [3]:
def is_prime(n):
    for i in range(2, n//2):
        if n % i == 0:
            return False
    return True

def filter_sums(MIN_SUM, MAX_SUM):
    sums = []
    for i in range(MIN_SUM, MAX_SUM):
        count = 0
        for addition in split_sum(i):
            if count_factors(addition[0] * addition[1]) <= 2:
                count += 1

            if (addition[0] > MAX_VALUE / 2 + 1 or addition[1] > MAX_VALUE / 2 + 1):
                if (is_prime(addition[0]) or is_prime(addition[1])):
                    return sums
            

        if (count == 0):
            sums.append(i)
    return sums 


MIN_SUM = 4 # 2+2 (мінімальна можлива комбінація чисел)
MAX_SUM = 198 # 99+99 (максимальна можлива комбінація чисел)
print("Всі можливі суми: ")
sums = filter_sums(MIN_SUM, MAX_SUM)
print(sums)

Всі можливі суми: 
[11, 17, 23, 27, 29, 35, 37, 41, 47, 53]


## 3. Добуток: "Тепер я знаю ці числа"

На даному етапі вчений, якому сказали добуток, повинен розкласти добуток на пари чисел, сума яких рівна одній із сум, які розрахував інший вчений. Також добутки, які зустрічаються більше одного разу мають бути видалені, оскільки якщо існує більше однієї пари чисел, які утворюють даний добуток і їхні суми дорівнюють одній із сум, які розрахував інший вчений, то цей вчений не зміг би точно назвати числа. Наприклад візьмемо добуток __30__. Цей добуток можуть утворити пари __(15, 2)__ і __(6, 5)__, вони утворюють суми __17__ і __11__, а ці числа співпадають з тими, що обрахував інший вчений. Це означає, що ці 2 пари чисел були б розв'язком задачі.

In [4]:
def filter_products(sums):
    products = []

    for sum in sums:
        s = set()
        for addition in split_sum(sum):
            s.add(addition[0] * addition[1])
        products.append(s)

    for i in range(len(products)):
        for j in range(i+1, len(products)):
            temp = products[i]
            products[i] = products[i] - products[j]
            products[j] = products[j] - temp
    
    return products

print("Всі можливі добутки: ")
products = filter_products(sums)
print(products)

Всі можливі добутки: 
[{24, 18, 28}, {52}, {112, 130, 76}, {162, 170, 140, 110, 176, 50, 182, 152, 92}, {100, 198, 168, 138, 204, 208, 54, 154, 190}, {96, 294, 234, 174, 304, 306, 276, 216, 250, 124}, {160, 232, 270, 336, 340, 186, 252}, {288, 418, 390, 364, 238, 400, 114, 148, 180, 310, 408, 378, 348, 414}, {480, 132, 550, 552, 522, 172, 496, 370, 532, 246, 280, 442, 540, 510}, {646, 520, 660, 282, 672, 682, 430, 690, 696, 570, 700, 702, 592, 612, 102, 360, 492, 240, 630}]


## 4. Сума: "Тепер я також знаю ці числа"

Тепер, для того щоб визначити числа, нам потрібно знайти суму, якій відповідає тільки один добуток. Як можна побачити вище, лише для суми __17__ відповідає тільки один добуток - __52__.

Тепер можна скласти рівняння:

$$ 
\begin{cases}a+b = 17\\ab = 52\end{cases}\Rightarrow\\
\begin{cases}b = 17 - a\\ab = 52\end{cases}\Rightarrow\\ 
$$

$$
a(17 - a) = 52\Rightarrow\\ a^2-17a+52=0\Rightarrow\\
$$

$$
D = b^2-4ac = 289 - 204 = 81\Rightarrow\\
$$

$$
a_1 = \frac{-b+\sqrt(D)}{2a} = \frac{17 + 9}{2} = 13\\
$$

$$
a_2 = \frac{-b-\sqrt(D)}{2a} = \frac{17  - 9}{2} = 4\\
$$

$$
b_1 = 17 - a_1 = 4\\
$$

$$
b_2 = 17 - a_2 = 13\\
$$

Числа __4__ і __13__ і є відповіддю до даної задачі

In [5]:
import math

def find_pairs(sums, products):
    for i in range(len(products)):
        if len(products[i]) == 1:
            D = sums[i]**2 - 4*list(products[i])[0]
            a1 = (sums[i] + math.sqrt(D)) / 2 
            a2 = (sums[i] - math.sqrt(D)) / 2
            b1 = sums[i] - a1
            b2 = sums[i] - a2

            print(f"a={a1}, b={b1}")
            print(f"a={a2}, b={b2}")

print("Знайдені пари чисел: ")
find_pairs(sums, products)

Знайдені пари чисел: 
a=13.0, b=4.0
a=4.0, b=13.0
